crud/batch_replace.lua

```diff
@@ -0,0 +1,246 @@
+        add_space_schema_hash = '?boolean',
+        fields = '?table',
+        stop_on_error = '?boolean',
+        rollback_on_error = '?boolean',
+        sharding_key_hash = '?number',
+        sharding_func_hash = '?number',
+        skip_sharding_hash_check = '?boolean',
+    })
+
+    opts = opts or {}
+
+    local space = box.space[space_name]
+    if space == nil then
+        return nil, {BatchReplaceError:new("Space %q doesn't exist", space_name)}
+    end
+
+    local _, err = sharding.check_sharding_hash(space_name,
+                                                opts.sharding_func_hash,
+                                                opts.sharding_key_hash,
+                                                opts.skip_sharding_hash_check)
+
+    if err ~= nil then
+        return nil, {err}
+    end
+
+    local inserted_tuples = {}
+    local errs = {}
+
+    box.begin()
+    for _, tuple in ipairs(tuples) do
+        -- add_space_schema_hash is true only in case of replace_object_many
+        -- the only one case when reloading schema can avoid replace error
+        -- is flattening object on router
+        local insert_result = schema.wrap_box_space_func_result(space, 'replace', {tuple}, {
+            add_space_schema_hash = opts.add_space_schema_hash,
+            field_names = opts.fields,
+        })
+
+        if insert_result.err ~= nil then
+            local err = {
+                err = insert_result.err,
+                tuple = tuple,
+            }
+
+            if opts.stop_on_error == true then
+                if opts.rollback_on_error == true then
+                    box.rollback()
+                    return nil, {err}
+                end
+
+                box.commit()
+
+                return inserted_tuples, {err}
+            end
+
+            table.insert(errs, err)
+        end
+
+        table.insert(inserted_tuples, insert_result.res)
+    end
+
+    if next(errs) ~= nil then
+        if opts.rollback_on_error == true then
+            box.rollback()
+            return nil, errs
+        end
+
+        box.commit()
+
+        return inserted_tuples, errs
+    end
+
+    box.commit()
+
+    return inserted_tuples
+end
+
+function batch_replace.init()
+    _G._crud.batch_replace_on_storage = batch_replace_on_storage
+end
+
+-- returns result, err, need_reload
+-- need_reload indicates if reloading schema could help
+-- see crud.common.schema.wrap_func_reload()
+local function call_batch_replace_on_router(space_name, original_tuples, opts)
+    dev_checks('string', 'table', {
+        timeout = '?number',
+        fields = '?table',
+        add_space_schema_hash = '?boolean',
+        stop_on_error = '?boolean',
+        rollback_on_error = '?boolean',
+    })
+
+    opts = opts or {}
+
+    local space = utils.get_space(space_name, vshard.router.routeall())
+    if space == nil then
+        return nil, {BatchReplaceError:new("Space %q doesn't exist", space_name)}, true
```

Comment:
We should add a couple of simple tests for sharding (and maybe schema) reload to cover these cases.

Reply:
Unfortunately reload sharding metadata/schema for batch operations turned out to be more difficult than originally expected, to recall request if sharding metadata/schema changed we should rollback batch operation on all storages, it is difficult, so for this point we should check sharding metadata/schema version before map call, it looks like separate PR. In that case we will be dealing with a latency.

Reply:
Well, let's support only the simplest case: an outdated router and updated storages (all of them). I don't think that it is ok to call crud requests under the migration

Reply:
You are right, now it is implemented.